File: itsm-prototype/src/lib/services/tag-service.ts
import { httpClient } from '@/lib/api/http-client';

export interface Tag {
  id: number;
  name: string;
  code: string;
  description?: string;
  color?: string;
  created_at?: string;
  updated_at?: string;
}

export interface CreateTagRequest {
  name: string;
  code: string;
  description?: string;
  color?: string;
}

class TagService {
  private readonly baseUrl = '/api/v1/tags';

  async listTags(): Promise<Tag[]> {
    return httpClient.get<Tag[]>(this.baseUrl);
  }

  async createTag(data: CreateTagRequest): Promise<Tag> {
    return httpClient.post<Tag>(this.baseUrl, data);
  }

  async updateTag(id: number, data: Partial<CreateTagRequest>): Promise<Tag> {
    return httpClient.put<Tag>(`${this.baseUrl}/${id}`, data);
  }

  async deleteTag(id: number): Promise<void> {
    return httpClient.delete(`${this.baseUrl}/${id}`);
  }

  async bindTag(tagId: number, entityType: string, entityId: number): Promise<void> {
    return httpClient.post<void>(`${this.baseUrl}/bind`, {
      tag_id: tagId,
      entity_type: entityType,
      entity_id: entityId,
    });
  }
}

export const tagService = new TagService();

File: itsm-backend/service/tag_service.go
package service

import (
	"context"
	"fmt"
	
	"itsm-backend/ent"
	"itsm-backend/ent/tag"
)

type TagService struct {
	client *ent.Client
}

func NewTagService(client *ent.Client) *TagService {
	return &TagService{client: client}
}

// CreateTag 创建通用标签
func (s *TagService) CreateTag(ctx context.Context, name, code, description, color string, tenantID int) (*ent.Tag, error) {
	return s.client.Tag.Create().
		SetName(name).
		SetCode(code).
		SetDescription(description).
		SetColor(color).
		SetTenantID(tenantID).
		Save(ctx)
}

// ListTags 获取标签列表
func (s *TagService) ListTags(ctx context.Context, tenantID int) ([]*ent.Tag, error) {
	return s.client.Tag.Query().
		Where(tag.TenantID(tenantID)).
		All(ctx)
}

// BindTagToEntity 为实体绑定标签
func (s *TagService) BindTagToEntity(ctx context.Context, tagID int, entityType string, entityID int) error {
	tagClient := s.client.Tag
	update := tagClient.UpdateOneID(tagID)

	switch entityType {
	case "project":
		update.AddProjectIDs(entityID)
	case "application":
		update.AddApplicationIDs(entityID)
	case "microservice":
		update.AddMicroserviceIDs(entityID)
	case "department":
		update.AddDepartmentIDs(entityID)
	case "team":
		update.AddTeamIDs(entityID)
	default:
		return fmt.Errorf("unsupported entity type: %s", entityType)
	}
	
	return update.Exec(ctx)
}

// UpdateTag 更新标签
func (s *TagService) UpdateTag(ctx context.Context, id int, name, code, description, color *string, tenantID int) (*ent.Tag, error) {
	// 检查标签是否存在且属于当前租户
	_, err := s.client.Tag.Query().
		Where(
			tag.IDEQ(id),
			tag.TenantIDEQ(tenantID),
		).
		First(ctx)
	if err != nil {
		return nil, err
	}

	update := s.client.Tag.UpdateOneID(id)
	if name != nil {
		update = update.SetName(*name)
	}
	if code != nil {
		update = update.SetCode(*code)
	}
	if description != nil {
		update = update.SetDescription(*description)
	}
	if color != nil {
		update = update.SetColor(*color)
	}

	return update.Save(ctx)
}

// DeleteTag 删除标签
func (s *TagService) DeleteTag(ctx context.Context, id int, tenantID int) error {
	// 检查标签是否存在且属于当前租户
	exists, err := s.client.Tag.Query().
		Where(
			tag.IDEQ(id),
			tag.TenantIDEQ(tenantID),
		).
		Exist(ctx)
	if err != nil {
		return err
	}
	if !exists {
		return fmt.Errorf("标签不存在: id=%d", id)
	}

	// 删除标签
	return s.client.Tag.DeleteOneID(id).Exec(ctx)
}
